### PR TITLE
Enable SEV backend

### PR DESCRIPTION
**Pull request by connorkuehl**

### CODE REVIEW

File: src/backend/sev/mod.rs [hunk 1]
```diff
@@ -193,7 +194,12 @@ impl backend::Backend for Backend {
...
 
-    fn build(&self, _code: Component, _sock: Option<&Path>) -> Result<Arc<dyn Keep>> {
-        unimplemented!()
+    fn build(&self, code: Component, _sock: Option<&Path>) -> Result<Arc<dyn Keep>> {
+        let shim = Component::from_bytes(SHIM)?;
+
+        let vm = Builder::new(Config::default(), shim, code, builder::Sev).build()?;
+        let keep = keep::Keep::new(Arc::new(RwLock::new(vm)))?;
+
+        Ok(Arc::new(keep))
```
Comment (haraldh) [inline src/backend/sev/mod.rs:203]: A lot of Arc

**Review by haraldh — APPROVED** [review]

File: src/backend/sev/builder.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+pub struct Sev;
+
+impl kvm::Hook for Sev {
+    fn code_loaded(&mut self, vm: &mut VmFd, addr_space: &[u8]) -> Result<()> {
+        use codicon::Decoder;
+
+        let mut sev = Firmware::open()?;
+        let build = sev.platform_status().unwrap().build;
+
+        let chain = env::var_os("SEV_CHAIN").ok_or_else(|| Error::from(ErrorKind::NotFound))?;
```
Comment (npmccallum) [inline src/backend/sev/builder.rs:28]: Please file a debt issue to improve the workflow around the SEV chain.

File: src/backend/sev/builder.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        let policy = Policy::default();
+        let session = Session::try_from(policy)?;
+        let start = session.start(chain)?;
+        let mut session = session.measure()?;
+        session.update_data(addr_space)?;
+
+        let (launcher, measurement) = {
+            let launcher = Launcher::new(vm, &mut sev)?;
+            let mut launcher = launcher.start(start)?;
+            launcher.update_data(addr_space)?;
```
Comment (npmccallum) [inline src/backend/sev/builder.rs:42]: I'm not convinced that this is correct. Please file an issue to review memory layout with me after the demo.

File: src/backend/sev/mod.rs [hunk 4]
```diff
@@ -193,7 +192,11 @@ impl backend::Backend for Backend {
...
     }
 
-    fn build(&self, _code: Component, _sock: Option<&Path>) -> Result<Arc<dyn Keep>> {
-        unimplemented!()
+    fn build(&self, code: Component, _sock: Option<&Path>) -> Result<Arc<dyn Keep>> {
+        let shim = Component::from_bytes(SHIM)?;
+
+        let vm = Builder::new(Config::default(), shim, code, builder::Sev).build()?;
+
+        Ok(Arc::new(RwLock::new(vm)))
```
Comment (npmccallum) [inline src/backend/sev/mod.rs:200]: Does this end up double-arcing the kvm?

Comment (connorkuehl) [reply src/backend/sev/mod.rs:200]: No. `Builder::build()` just returns a bare `struct Vm`, so this `Arc::new()` is the first and only `Arc`.

**Review by npmccallum — APPROVED** [review]

**Review by haraldh — APPROVED** [review]

**Pull request merged**